File: Linear Control Course/Chapter17/Lesson3/Chapter17_Lesson3.cpp
#include <iostream>
#include <cmath>

// Compute delay margin from phase margin (deg) and gain crossover (rad/s)
double delayMargin(double phaseMarginDeg, double w_gc) {
    const double phaseMarginRad = phaseMarginDeg * M_PI / 180.0;
    return phaseMarginRad / w_gc;
}

int main() {
    double phaseMarginDeg = 45.0;   // e.g., from offline Bode plot
    double w_gc = 5.0;              // rad/s
    double T_d = delayMargin(phaseMarginDeg, w_gc);
    std::cout << "Delay margin (s): " << T_d << std::endl;

    // In a ROS node, this function can be used to check whether the estimated
    // communication + computation delay stays below T_d for a given controller.
    return 0;
}

File: Linear_Control_Course/Chapter17/Lesson3/Chapter17_Lesson3.java
public class DelayMarginExample {

    // Compute delay margin from phase margin (deg) and gain crossover (rad/s)
    public static double delayMargin(double phaseMarginDeg, double wGc) {
        double phaseMarginRad = Math.toRadians(phaseMarginDeg);
        return phaseMarginRad / wGc;
    }

    public static void main(String[] args) {
        double phaseMarginDeg = 45.0; // from design tool or measurement
        double wGc = 5.0;             // rad/s
        double Td = delayMargin(phaseMarginDeg, wGc);
        System.out.println("Delay margin (s): " + Td);

        // In a robotics framework such as WPILib, this can guide the selection
        // of sampling times and network delays allowed in a closed-loop system.
    }
}
